downloadDirectory/src/main/java/org/reactome/release/downloadDirectory/GenerateGOAnnotationFile.java

```diff
@@ -0,0 +1,413 @@
+                    // TODO: Rule explanation
+                    if (goTerm.equals(CELLULAR_COMPONENT) || (goTerm.equals(BIOLOGICAL_PROCESS) && catalystInstances.size() == 0)) {
+                        Set<GKInstance> proteinInstances = getReactionProteins(reactionInst);
+                        processProteins(goTerm, proteinInstances, reactionInst, null);
+                    } else {
+                        // If looking at MF of an instance, or BP of an instance that has catalyst instances.
+                        for (GKInstance catalystInst : catalystInstances) {
+                            // Check that catalyst instance has no disqualifying attributes.
+                            if (validCatalyst(catalystInst, goTerm)) {
+                                Set<GKInstance> proteinInstances = new HashSet<>();
+                                // For MF annotations, gets all proteins from the ActiveUnit if it exists or from the PhysicalEntity if not.
+                                if (goTerm.equals(MOLECULAR_FUNCTION)) {
+                                    GKInstance activeUnitInst = (GKInstance) catalystInst.getAttributeValue(ReactomeJavaConstants.activeUnit);
+                                    GKInstance entityInst = activeUnitInst == null ? (GKInstance) catalystInst.getAttributeValue(ReactomeJavaConstants.physicalEntity) : activeUnitInst;
+                                    proteinInstances = getMolecularFunctionProteins(entityInst);
+                                    // For BP annotations (that have a catalyst), it gets all proteins from the PhysicalEntity
+                                } else if (goTerm.equals(BIOLOGICAL_PROCESS)) {
+                                    proteinInstances = getBiologicalProcessProteins((GKInstance) catalystInst.getAttributeValue(ReactomeJavaConstants.physicalEntity));
+                                }
+                                processProteins(goTerm, proteinInstances, reactionInst, catalystInst);
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        outputGOAFile();
+        Files.move(Paths.get(GOA_FILENAME), Paths.get(releaseNumber + "/" + GOA_FILENAME), StandardCopyOption.REPLACE_EXISTING);
+        logger.info("Finished generating gene_association.reactome");
+    }
+
+    // This method uses an AttributeQueryRequest to find all Proteins (above and below in the hierarchy) associated with a ReactionlikeEvent.
+    private static Set<GKInstance> getReactionProteins(GKInstance reactionInst) throws Exception {
+        List<ClassAttributeFollowingInstruction> classesToFollow = new ArrayList<>();
+        classesToFollow.add(new ClassAttributeFollowingInstruction(ReactomeJavaConstants.Pathway, new String[]{ReactomeJavaConstants.hasEvent}, new String[]{}));
```

Comment:
Maybe this can be organized as an ordered map in a structure like
 `Map<String, List<String>> classToAttributes = new LinkedHashMap<>();` 
so that the common code to build the instruction and add to `classesToFollow` can be a single line in a loop going through the entries?